### PR TITLE
Adds delta/dictionary encoding for ints.

### PR DESCRIPTION
**Pull request by notbdu**
Adds an implementation for the int encoder/decoder. Implemented delta and dictionary encoding.

### CODE REVIEW

File: encoding/int_encode.go [hunk 1]
```diff
@@ -1,30 +1,188 @@
...
+		// Table encode if there is less than 256 unique valuesIt.
+		enc.metaProto.Encoding = encodingpb.EncodingType_DICTIONARY
+		// Min number of bits to encode each actual value into the table.
+		enc.metaProto.BytesPerDictionaryValue = int64(math.Ceil(float64(bits.Len(uint(max))) / 8.0))
+		// Min number of bits required to encode table indices.
+		enc.metaProto.BitsPerEncodedValue = int64(bits.Len(uint(len(dict))))
+	default:
+		// Default to delta encoding.
+		enc.metaProto.Encoding = encodingpb.EncodingType_DELTA
+		enc.metaProto.BitsPerEncodedValue = int64(bits.Len(uint(max - min)))
```
Comment (xichen2020) [inline encoding/int_encode.go]: Probably should be `max - min + 1` since you need to encode all values in [`min`, `max`].

File: encoding/int_encode.go [hunk 2]
```diff
@@ -1,30 +1,188 @@
...
+	enc.dictionaryProto.Reset()
+}
+
+func (enc *IntEnc) encodeDelta(
+	writer io.Writer,
+	bitsPerEncodedValue int64,
+	valuesIt RewindableIntIterator,
+) error {
+	// Get first record and set as the start.
+	valuesIt.Next()
```
Comment (xichen2020) [inline encoding/int_encode.go]: Handle the case where there's no values?

File: encoding/int_encode.go [hunk 3]
```diff
@@ -1,30 +1,188 @@
...
+	}
+
+	var delta int
+	last := curr
+	for valuesIt.Next() {
+		curr = valuesIt.Current()
+		delta = curr - last
+		if delta < 0 {
+			// Flip the sign.
+			delta *= -1
```
Comment (xichen2020) [inline encoding/int_encode.go]: `delta = -delta` is probably more efficient here.

File: encoding/int_encode.go [hunk 4]
```diff
@@ -1,30 +1,188 @@
...
+			// Set the MSB if the sign is negative.
+			delta |= 1 << uint(bitsPerEncodedValue)
```
Comment (xichen2020) [inline encoding/int_encode.go]: Move `1 << uint(bitsPerEncodedValue)` outside the for loop so it's only calculated once.

File: encoding/int_encode.go [hunk 5]
```diff
@@ -1,30 +1,188 @@
...
+	dict map[int]int,
+	valuesIt RewindableIntIterator,
+) error {
+	sortedDict := make([]uint64, len(dict))
+	for value, idx := range dict {
+		sortedDict[idx] = uint64(value)
+	}
+
+	var start int
+	enc.dictionaryProto.Data = make([]byte, len(sortedDict)*int(bytesPerDictionaryValue))
```
Comment (xichen2020) [inline encoding/int_encode.go]: This should be reused across encoding calls.

File: encoding/int_encode.go [hunk 6]
```diff
@@ -1,30 +1,188 @@
...
+		start = idx * int(bytesPerDictionaryValue)
+		copy(enc.dictionaryProto.Data[start:start+int(bytesPerDictionaryValue)], enc.buf[:bytesPerDictionaryValue])
+	}
+
+	if err := proto.EncodeIntDictionary(&enc.dictionaryProto, &enc.buf, writer); err != nil {
+		return err
+	}
+
+	bitWriter := bitstream.NewWriter(writer)
+	var encodedValue uint64
```
Comment (xichen2020) [inline encoding/int_encode.go]: No need to declare this outside the loop, can declare where its assigned.

File: encoding/dictionary_based_int_iterator.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/xichen2020/eventdb/generated/proto/encodingpb"
+	"github.com/xichen2020/eventdb/x/io"
+	"github.com/xichen2020/eventdb/x/proto"
+
+	bitstream "github.com/dgryski/go-bitstream"
+)
+
+// DictionaryBasedIntIterator iterates through a dict encoded stream of ints.
+type DictionaryBasedIntIterator struct {
+	bitReader               *bitstream.BitReader
```
Comment (xichen2020) [inline encoding/dictionary_based_int_iterator.go]: Should probably make this a field of the decoder, and call `bitReader.Reset(reader)` for every decode, and pass in the bitreader to the iterator constructor, so you don't need to create a new bit reader every time.

File: encoding/int_encode.go [hunk 8]
```diff
@@ -1,30 +1,188 @@
...
+		endianness.PutUint64(enc.buf, value)
+		start = idx * int(bytesPerDictionaryValue)
+		copy(enc.dictionaryProto.Data[start:start+int(bytesPerDictionaryValue)], enc.buf[:bytesPerDictionaryValue])
+	}
+
+	if err := proto.EncodeIntDictionary(&enc.dictionaryProto, &enc.buf, writer); err != nil {
+		return err
+	}
+
+	bitWriter := bitstream.NewWriter(writer)
```
Comment (xichen2020) [inline encoding/int_encode.go]: Should probably make this a field of the encoder, and call `bitWriter.Reset(writer)` for every encode, and reuse the bitwriter here, so you don't need to create a new bit writer every time.

File: encoding/delta_int_iterator.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+		d.curr -= int(delta)
+	} else {
+		d.curr += int(delta)
+	}
+
+	return true
+}
+
+// Current returns the current int.
+func (d *DeltaIntIterator) Current() int {
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/delta_int_iterator.go]: 👍

File: encoding/delta_int_iterator.go [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+	// Read in an extra bit for the sign.
+	var (
+		delta uint64
+	)
+	delta, d.err = d.bitReader.ReadBits(int(d.bitsPerEncodedValue) + 1)
+	if d.err != nil {
+		return false
+	}
+	// Check if negative bit is set.
+	isNegative := delta&d.negativeBit == d.negativeBit
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/delta_int_iterator.go]: I'll check this.

File: encoding/delta_int_iterator.go [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+	return d.curr
+}
+
+// Err returns any error recorded while iterating.
+func (d *DeltaIntIterator) Err() error {
+	return d.err
+}
+
+// Close the iterator.
+func (d *DeltaIntIterator) Close() error {
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/delta_int_iterator.go]: 👍

File: encoding/int_encode.go [hunk 12]
```diff
@@ -1,30 +1,188 @@
...
+	for value, idx := range dict {
+		sortedDict[idx] = uint64(value)
+	}
+
+	var start int
+	enc.dictionaryProto.Data = make([]byte, len(sortedDict)*int(bytesPerDictionaryValue))
+	for idx, value := range sortedDict {
+		endianness.PutUint64(enc.buf, value)
+		start = idx * int(bytesPerDictionaryValue)
+		copy(enc.dictionaryProto.Data[start:start+int(bytesPerDictionaryValue)], enc.buf[:bytesPerDictionaryValue])
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: The bit reader/writer impl doesn't have a `Seek` method. I'd have to fork that impl and add it. We could do manual seeking outside of that lib but that would require figuring out the byte offset, instantiating a new bitreader for sliced from that offset and reading in bits up until the bit offset that we wanted to seek to.

File: encoding/dictionary_based_int_iterator.go [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+	bytesPerDictionaryValue int64
+	bitsPerEncodedValue     int64
+	extBuf                  *[]byte
+	dict                    []byte
+	curr                    int
+	err                     error
+	closed                  bool
+}
+
+// NewDictionaryBasedIntIterator returns a new dictionary based int iterator.
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/dictionary_based_int_iterator.go]: 👍

File: encoding/dictionary_based_int_iterator.go [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+	// Read the idx into the dict first.
+	var dictIdx uint64
+	dictIdx, d.err = d.bitReader.ReadBits(int(d.bitsPerEncodedValue))
+	if d.err != nil {
+		return false
+	}
+
+	// Use idx to fetch value.
+	start := int64(dictIdx) * d.bytesPerDictionaryValue
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/dictionary_based_int_iterator.go]: Added a comment.

File: encoding/dictionary_based_int_iterator.go [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+	copy((*d.extBuf)[:d.bytesPerDictionaryValue], d.dict[start:start+d.bytesPerDictionaryValue])
+	d.curr = int(endianness.Uint64(*d.extBuf))
+	return true
+}
+
+// Current returns the current int.
+func (d *DictionaryBasedIntIterator) Current() int {
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/dictionary_based_int_iterator.go]: 👍

File: encoding/dictionary_based_int_iterator.go [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// Err returns any error recorded while iterating.
+func (d *DictionaryBasedIntIterator) Err() error {
+	return d.err
+}
+
+// Close the iterator.
+func (d *DictionaryBasedIntIterator) Close() error {
+	d.closed = true
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/dictionary_based_int_iterator.go]: 👍

File: encoding/encoding.go [hunk 17]
```diff
@@ -1,11 +1,15 @@
...
 )
 
-// Reader is both an io.Reader and an io.ByteReader.
-type Reader interface {
-	io.Reader
-	io.ByteReader
-}
+// ByteOrder for data serialization.
+var (
+	endianness = binary.LittleEndian
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/encoding.go]: 👍

File: encoding/int_decode.go [hunk 18]
```diff
@@ -1,10 +1,75 @@
...
+		iter, err = dec.decodeDictionary(reader, dec.metaProto.BytesPerDictionaryValue, dec.metaProto.BitsPerEncodedValue)
+	default:
+		return nil, fmt.Errorf("Invalid encoding type: %v", dec.metaProto.Encoding)
+	}
+
+	return iter, err
+}
+
+// Reset the int decoder.
+func (dec *IntDec) Reset() {}
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_decode.go]: Added some cleanup here.

File: encoding/int_encode.go [hunk 19]
```diff
@@ -1,30 +1,188 @@
...
 
 // Encode encodes a collection of ints and writes the encoded bytes to the writer.
-func (enc *IntEnc) Encode(writer io.Writer, values ForwardIntIterator) error {
-	return errors.New("not implemented")
+func (enc *IntEnc) Encode(
+	writer io.Writer,
+	valuesIt RewindableIntIterator,
+) error {
+	// Determine whether we want to do table compression
+	var (
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/int_encode.go [hunk 20]
```diff
@@ -1,30 +1,188 @@
...
+
+	// Rewind iteration.
+	valuesIt.Rewind()
+
+	switch {
+	case len(dict) <= dictEncodingMaxCardinalityInt:
+		// Table encode if there is less than 256 unique valuesIt.
+		enc.metaProto.Encoding = encodingpb.EncodingType_DICTIONARY
+		// Min number of bits to encode each actual value into the table.
+		enc.metaProto.BytesPerDictionaryValue = int64(math.Ceil(float64(bits.Len(uint(max))) / 8.0))
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/int_encode.go (hunk 1 above)
Comment (notbdu) [reply encoding/int_encode.go]: I only need to encode the max possible delta which is `max - min`.

File: encoding/int_encode.go (hunk 2 above)
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/int_encode.go (hunk 3 above)
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/int_encode.go (hunk 4 above)
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/int_encode.go (hunk 5 above)
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/int_encode.go (hunk 6 above)
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/dictionary_based_int_iterator.go (hunk 7 above)
Comment (notbdu) [reply encoding/dictionary_based_int_iterator.go]: 👍

File: encoding/int_encode.go (hunk 8 above)
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/int_encode.go [hunk 21]
```diff
@@ -1,30 +1,188 @@
...
+}
 
 // Encode encodes a collection of ints and writes the encoded bytes to the writer.
-func (enc *IntEnc) Encode(writer io.Writer, values ForwardIntIterator) error {
-	return errors.New("not implemented")
+func (enc *IntEnc) Encode(
+	writer io.Writer,
+	valuesIt RewindableIntIterator,
+) error {
+	// Determine whether we want to do table compression
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/delta_int_iterator.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/delta_int_iterator.go]: Looks like it works w/o the parens. Both sides are prob evaluated separately before evaluating the `==`.

File: encoding/delta_int_iterator.go [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+	curr                int
+	err                 error
+	closed              bool
+}
+
+// NewDeltaIntIterator returns a new delta encoded int iterator.
+func NewDeltaIntIterator(
+	reader io.Reader,
+	bitsPerEncodedValue int64,
+	deltaStart int64,
```
[the remark this answers is not in the document]
Comment (xichen2020) [reply encoding/delta_int_iterator.go]: Sounds good.

File: encoding/delta_int_iterator.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (xichen2020) [reply encoding/delta_int_iterator.go]: Ok thanks, perhaps still put a pair of parens around it for better readability.

File: encoding/int_encode.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (xichen2020) [reply encoding/int_encode.go]: Hm, can you explain the scenario where seeking is needed?

File: encoding/dictionary_based_int_iterator.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (xichen2020) [reply encoding/dictionary_based_int_iterator.go]: On a second thought, I'm more in favor of simply implementing this more explicitly in `x/io` probably and not using endianness at all, something along the lines of:
```
// Precondition: n <= 8 && len(buf) >= n
func writeInt(x int64, n int, buf []byte) {
  for i := 0; i < n; i++ {
    buf[i] = x &0xFF
    x >>= 8
  }
}

// Precondition: n <= 8 && len(buf) >= n
func readInt(n int, buf []byte) int64 {
  var x int64
  for i := 0; i < n; i++ {
    x <<= 8
    x |= buf[i]
  }
  return x
}
```

File: encoding/int_encode.go (hunk 1 above)
Comment (xichen2020) [reply encoding/int_encode.go]: Sure, but you'd need to encode all values between [0, max - min], and the total count of that is still `max - min + 1` values.

File: encoding/int_encode.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: To read N bits @ idx offset IDX, we would need to seek to N * IDX bits first and then read out N bits. 

Actually, I would probably implement a `ReadAt` method to read N bits @ bit offset Y.

[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: I have a TODO for this to be done later.

[the remark this answers is not in the document]
Comment (xichen2020) [reply encoding/int_encode.go]: Yeah I was more asking "when will we need to directly read N bits @ idx offset without reading the bits before"?

[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: When the dict is used to lookup the actual value. The encoded value is the idx into the dictionary. So in the decode path, we'll need to read the actual value given an idx.

File: encoding/dictionary_based_int_iterator.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/dictionary_based_int_iterator.go]: 👍

File: encoding/int_encode.go (hunk 1 above)
Comment (notbdu) [reply encoding/int_encode.go]: Mmmm well if there are enough bits to represent the value of the largest delta (`max - min`) aren't there enough bits to represent every other value from that number to 0? Adding an extra bit is unnecessary here. 

Consider the simple case where:
```
max := 7
min := 4
numBits := bits.Len(uint(max - min))
log.Println("numBits:", numBits)
log.Printf("%b", 3)
log.Printf("%b", 2)
log.Printf("%b", 1)
log.Printf("%b", 0)

--- STDOUT ---
2018/12/09 20:01:22 numBits: 2
2018/12/09 20:01:22 11
2018/12/09 20:01:22 10
2018/12/09 20:01:22 1
2018/12/09 20:01:22 0
```

As long as there are enough bits to represent `max-min`, then there are enough bits to represent all other numbers below that as well.

File: encoding/delta_int_iterator.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/delta_int_iterator.go]: 👍

File: encoding/int_encode.go (hunk 1 above)
Comment (xichen2020) [reply encoding/int_encode.go]: Yeah it seems `bits.Len(n)` returns the number of bits needed to represent numbers in [0,n] (i.e., n+1 numbers), so we are talking about the same thing here.

**Comment by xichen2020** [comment]
Seems metalint is still failing though?

File: encoding/int_encode.go (hunk 1 above)
Comment (notbdu) [reply encoding/int_encode.go]: Ahh kk, 👍.

**Comment by notbdu** [comment]
@xichen2020 fixed!

**Comment by xichen2020** [comment]
Nice, will have another look.

File: encoding/int_decode.go [hunk 23]
```diff
@@ -1,10 +1,90 @@
...
+		iter, err = dec.decodeDictionary(reader)
+	default:
+		return nil, fmt.Errorf("Invalid encoding type: %v", dec.metaProto.Encoding)
+	}
+
+	return iter, err
+}
+
+// Reset the int decoder.
+func (dec *IntDec) Reset() {
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_decode.go]: :thumb:

File: encoding/delta_int_iterator.go [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+	return true
+}
+
+// Current returns the current int.
+func (d *DeltaIntIterator) Current() int {
+	return d.curr
+}
+
+// Err returns any error recorded while iterating.
+func (d *DeltaIntIterator) Err() error {
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/delta_int_iterator.go]: 👍

File: encoding/delta_int_iterator.go [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+func (d *DeltaIntIterator) Next() bool {
+	if d.closed || d.err != nil {
+		return false
+	}
+
+	// Read in an extra bit for the sign.
+	var (
+		delta uint64
+	)
+	delta, d.err = d.bitReader.ReadBits(int(d.bitsPerEncodedValue) + 1)
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/delta_int_iterator.go]: 👍

File: encoding/delta_int_iterator.go [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+	bitReader           *bitstream.BitReader
+	bitsPerEncodedValue int64
+	negativeBit         uint64
+	curr                int
+	err                 error
+	closed              bool
+}
+
+// NewDeltaIntIterator returns a new delta encoded int iterator.
+func NewDeltaIntIterator(
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/delta_int_iterator.go]: Nope, making all of these iterator ctors private.

File: encoding/dictionary_based_int_iterator.go [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+	bitsPerEncodedValue     int64
+	extBuf                  *[]byte
+	dict                    []byte
+	curr                    int
+	err                     error
+	closed                  bool
+}
+
+// NewDictionaryBasedIntIterator returns a new dictionary based int iterator.
+func NewDictionaryBasedIntIterator(
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/dictionary_based_int_iterator.go]: Same comment as above.

File: encoding/dictionary_based_int_iterator.go [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+	extBuf *[]byte, // extBuf is an external byte buffer for memory re-use.
+	minValue int64,
+	bytesPerDictionaryValue int64,
+	bitsPerEncodedValue int64,
+) (*DictionaryBasedIntIterator, error) {
+	if err := proto.DecodeIntDictionary(extProto, extBuf, reader); err != nil {
+		return nil, err
+	}
+	// Zero out extBuf so we can re-use it during iteration.
+	io.WriteInt(uint64(0), 8, *extBuf)
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/dictionary_based_int_iterator.go]: Oh yah, not anymore since we have our own read/write int methods.

File: encoding/int_decode.go [hunk 29]
```diff
@@ -1,10 +1,90 @@
 package encoding
 
+import (
+	"bytes"
+	"fmt"
+
+	bitstream "github.com/dgryski/go-bitstream"
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_decode.go]: 👍

File: encoding/int_decode.go [hunk 30]
```diff
@@ -1,10 +1,90 @@
...
+	buf             []byte
+}
+
+// NewIntDecoder creates a new int Decoder.
+func NewIntDecoder() *IntDec {
+	return &IntDec{
+		// Make buf at least big enough to hold Uint64 values.
+		buf: make([]byte, uint64SizeBytes),
+		// Make a bitWriter w/ an empty read buffer that will be re-used for every `Decode` call.
+		bitReader: bitstream.NewReader(&bytes.Buffer{}),
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_decode.go]: 👍

File: encoding/int_encode.go [hunk 31]
```diff
@@ -1,30 +1,216 @@
...
+		// This way we track if we've exceeded the max # of uniques.
+		if len(dict) <= dictEncodingMaxCardinalityInt {
+			if _, ok := dict[curr]; !ok {
+				dict[curr] = idx
+				idx++
+			}
+		}
+
+		// Track max/min.
+		max = int(math.Max(float64(curr), float64(max)))
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/int_encode.go [hunk 32]
```diff
@@ -1,30 +1,216 @@
...
+	}
+
+	if valuesIt.Err() != nil {
+		return valuesIt.Err()
+	}
+
+	// Rewind iteration.
+	valuesIt.Rewind()
+
+	switch {
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/int_encode.go [hunk 33]
```diff
@@ -1,30 +1,216 @@
...
+		max  = math.MinInt64
+		min  = math.MaxInt64
+		idx  int
+		dict = make(map[int]int, dictEncodingMaxCardinalityInt)
+	)
+	for valuesIt.Next() {
+		curr := valuesIt.Current()
+
+		if idx == 0 {
+			enc.metaProto.DeltaStart = int64(curr)
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/int_encode.go [hunk 34]
```diff
@@ -1,30 +1,216 @@
...
+	valuesIt RewindableIntIterator,
+) error {
+	// Encode the first value which is always a delta of 0.
+	if !valuesIt.Next() {
+		return valuesIt.Err()
+	}
+
+	// Sanity check.
+	curr := valuesIt.Current()
+	if curr != int(enc.metaProto.DeltaStart) {
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: Yea it shouldn't be false. I can remove the sanity checks.

File: encoding/int_encode.go [hunk 35]
```diff
@@ -1,30 +1,216 @@
...
+		return fmt.Errorf("first value (%d) and encoded delta start value (%d) must match", curr, enc.metaProto.DeltaStart)
+	}
+
+	// Write an extra bit to encode the sign of the delta.
+	if err := enc.bitWriter.WriteBits(uint64(0), int(bitsPerEncodedValue)+1); err != nil {
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: encoding/int_encode.go [hunk 36]
```diff
@@ -1,30 +1,216 @@
...
+	enc.dictionaryProtoData = xbytes.EnsureBufferSize(enc.dictionaryProtoData, dictionaryProtoSize, xbytes.DontCopyData)
+	for idx, value := range sortedDict {
+		// The dictionary value is encoded as a positive number to be added to the minimum value.
+		dictValue := value - int(enc.metaProto.MinValue)
+		// Sanity check.
+		if dictValue < 0 {
+			return fmt.Errorf("dictionary values (%d) should not be less than 0", dictValue)
+		}
+		xio.WriteInt(uint64(dictValue), int(bytesPerDictionaryValue), enc.buf)
+		start := idx * int(bytesPerDictionaryValue)
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: 👍

File: generated/proto/encodingpb/int.proto [hunk 37]
```diff
@@ -0,0 +1,33 @@
...
+
+option go_package = "github.com/xichen2020/eventdb/generated/proto/encodingpb";
+
+import "github.com/xichen2020/eventdb/generated/proto/encodingpb/types.proto";
+
+package encodingpb;
+
+message IntMeta {
+    EncodingType encoding = 1;
+    int64 num_records = 2;
```
[the remark this answers is not in the document]
Comment (notbdu) [reply generated/proto/encodingpb/int.proto]: Removed this actually. Wasn't being used anywhere.

File: x/io/io.go [hunk 38]
```diff
@@ -0,0 +1,21 @@
+package io
```
[the remark this answers is not in the document]
Comment (notbdu) [reply x/io/io.go]: 👍

File: generated/proto/encodingpb/int.proto [hunk 39]
```diff
@@ -0,0 +1,33 @@
...
+    // Bytes per dictionary value since we can't do bit packing 
+    // in the int dictionary yet. See TODO below.
+    int64 bytes_per_dictionary_value = 3;
```
[the remark this answers is not in the document]
Comment (notbdu) [reply generated/proto/encodingpb/int.proto]: Probably better to leave as is since there are > 1 fields required (e.g. `bytes_per_dictionary_value` and `min_value` for dict encoding). Can't use `oneof` w/o further grouping the fields into their own messages.

[the remark this answers is not in the document]
Comment (xichen2020) [reply generated/proto/encodingpb/int.proto]: Yeah I meant grouping the fields needed for dictionary encoding and those for delta encoding as different message types, and use `oneof` for them. Take it or leave it though.

**Comment by notbdu** [comment]
Will run `glide up` later, runs horribly slow on the plane...

File: encoding/int_encode.go [hunk 40]
```diff
@@ -1,30 +1,216 @@
...
+	if !valuesIt.Next() {
+		return valuesIt.Err()
+	}
+
+	// Write an extra bit to encode the sign of the delta.
+	if err := enc.bitWriter.WriteBits(uint64(0), int(bitsPerEncodedValue)); err != nil {
+		return err
+	}
+
+	negativeBit := 1 << uint(bitsPerEncodedValue)
```
Comment (xichen2020) [inline encoding/int_encode.go]: Hm should this be `bitsPerEncodedValue - 1` now? Guess we don't have a unit test to catch this?

Comment (notbdu) [reply encoding/int_encode.go]: Adding unit tests for negative numbers too.

File: encoding/int_encode.go [hunk 41]
```diff
@@ -196,7 +194,9 @@ func (enc *IntEnc) encodeDictionary(
 			return fmt.Errorf("dictionary values (%d) should not be less than 0", dictValue)
 		}
 		xio.WriteInt(uint64(dictValue), int(bytesPerDictionaryValue), enc.buf)
-		start := idx * int(bytesPerDictionaryValue)
+		if idx > 0 {
```
[the remark this answers is not in the document]
Comment (notbdu) [reply encoding/int_encode.go]: Ah, right XD.

File: persist/fs/writer.go [hunk 42]
```diff
@@ -365,13 +365,11 @@ func (w *writer) writeValues(
 		w.bw.Reset()
 		return w.bw.Encode(writer, valueIt.boolIt)
 	case field.IntType:
-		w.iw.Reset()
 		return w.iw.Encode(writer, valueIt.intIt)
 	case field.DoubleType:
 		w.dw.Reset()
```
Comment (xichen2020) [inline persist/fs/writer.go]: I imagine you need to remove the `Reset()` for other types as well?

Comment (xichen2020) [reply persist/fs/writer.go]: And from their encoding APIs, too.

**Comment by notbdu** [comment]
Yea.... so for some reason I kept thinking I’d do it in the type specific
PRs to come. But I guess that doesn’t make as much sense haha.
